### ISSUE TITLE
Kafka receiver stuck while shutting down at v0.93.0

### ISSUE THREAD

**Issue by james-ryans** [open-telemetry/opentelemetry-collector-contrib#30789]
### Component(s)

receiver/kafka

## Description
Shutting down kafka receiver got stuck forever while transitioning from `StatusStopping` to `StatusStopped`.

I've debugged this for a while and apparently this is because of `consumeLoop` returns `context canceled` error and do `ReportStatus FatalErrorEvent` at `receiver/kafkareceiver/kafka_receiver.go` between line 163-165. But the `sync.Mutex.Lock()` in `ReportStatus FatalErrorEvent` never gets unlocked (I don't know why), so that the `ReportStatus` for `StatusStopped` stuck forever while trying to acquire the mutex lock.

Also, I've tried to rollback to v0.92.0 and it works well. And traced to issue down to `receiver/kafkareceiver/kafka_receiver.go` at line 164 `c.settings.ReportStatus(component.NewFatalErrorEvent(err))` changed at PR https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30593 was the cause.

## Steps to Reproduce
Create a collector with `kafkareceiver` factory in it. And have a `receivers.kafka` in the config.

## Expected Result
Should be able to shutdown properly.

## Actual Result
Stuck indefinitely while shutting down with the logs below.

### Collector version

v0.93.0

### OpenTelemetry Collector configuration

```yaml
service:
  pipelines:
    traces:
      receivers: [kafka]

receivers:
  kafka:
    brokers:
      - localhost:9092
    encoding: otlp_proto # available encodings are otlp_proto, jaeger_proto, jaeger_json, zipkin_proto, zipkin_json, zipkin_thrift
    initial_offset: earliest # consume messages from the beginning
```

### Log output

```shell
2024-01-26T08:32:45.266+0700	info	kafkareceiver@v0.93.0/kafka_receiver.go:431	Starting consumer group	{"kind": "receiver", "name": "kafka", "data_type": "traces", "partition": 0}
^C2024-01-26T08:32:53.626+0700	info	otelcol@v0.93.0/collector.go:258	Received signal from OS	{"signal": "interrupt"}
2024-01-26T08:32:53.626+0700	info	service@v0.93.0/service.go:179	Starting shutdown...
2024-01-26T08:32:54.010+0700	info	kafkareceiver@v0.93.0/kafka_receiver.go:181	Consumer stopped	{"kind": "receiver", "name": "kafka", "data_type": "traces", "error": "context canceled"}
```

**Comment by crobert-1** [comment]
@mwear: Do you have any thoughts on why a component may never be able to [get the lock when reporting status](https://github.com/open-telemetry/opentelemetry-collector/blob/9082cb16c1edb9aa44a717ec149a0791e59c9b78/service/internal/status/status.go#L140)? It looks like this may be related to the work you've been doing on component status reporting.

Possible related PR: https://github.com/open-telemetry/opentelemetry-collector/pull/8836

**Comment by mwear** [comment]
Based on the research @james-ryans did, this came in after this change: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30610. What I suspect is happening is that writing the fatal error to the asyncErrorChannel in serviceHost is blocking, so that ReportStatus never returns (and never releases its lock). Here is the suspect line: https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/host.go#L73.

I think this is a variation of this existing problem: https://github.com/open-telemetry/opentelemetry-collector/issues/8116, which is also assigned to me. It has been on my todo list. I'll look into it.

**Comment by crobert-1** [comment]
Thanks @mwear, appreciate your insight here!

**Comment by atoulme** [comment]
This is https://github.com/open-telemetry/opentelemetry-collector/issues/9824.

**Comment by crobert-1** [comment]
> > This is [open-telemetry/opentelemetry-collector#9824](https://github.com/open-telemetry/opentelemetry-collector/issues/9824).
> 
> Here is an example of a flaky test in the collector-contrib due to the same issue happening for the `opencensus` receiver: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8742859512/job/23992117763. The test has goroutine dump which indicates to the same problem with `asyncErrorChannel` as is pointed out in the shared issue.

Adding a reference to the issue for the flaky test: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27295

**Comment by crobert-1** [comment]
+1 freq: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32667

**Comment by Dennis8274** [comment]
quick fix like as follow?
<img width="857" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/20966346/2776a514-9a69-40f1-b5a5-90ff1bb59b48">

**Comment by Dennis8274** [comment]
quick fix as follows?
<img width="865" alt="image" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/20966346/e1451b74-86c8-43ae-9eba-607ae7c61774">

**Comment by crobert-1** [comment]
Your fixed worked @Dennis8274, thanks for the suggestion! I've posted a PR to resolve this issue. 👍

**Comment by tejas-contentstack** [comment]
I'm facing a similar issue while shutting down the Kafka receiver. If we try to shut down the collector, it stops the consumer however, starts the consumer right after. Check in the logs snapshot

Collector version: `v0.101.0`

Here's the Otel config file:

```
receivers:

  kafka/metrics:
    brokers:  ["0.0.0.0:9092", "localhost:9092", "localhost:9092"]
    group_id: otel-metrics-consumer
    topic: topic-otel-metrics
    header_extraction:
      extract_headers: true
      headers:
        - uid
    metadata:
      full: true
      
exporters:
  exporter:


service:
  telemetry:
    logs:
      output_paths: ["stdout"]
      error_output_paths: ["stderr"]
    metrics:
      level: none
  extensions: []
  pipelines: 
    metrics:
      receivers: [kafka/metrics]
      processors: []
      exporters: [exporter]
```


<img width="752" alt="Screenshot 2024-08-30 at 1 11 42 PM" src="https://github.com/user-attachments/assets/5955f4a3-d36d-4751-9b52-81cd368c6774">

**Comment by crobert-1** [comment]
Hello @tejas-contentstack, thanks for adding frequency! In this case it may be best to open a new issue and reference this one with it, since the error message is slightly different. I agree it looks like it may be a similar problem, but it would be best to investigate fully to make sure we're not missing anything 👍

**Comment by tbm48813** [comment]
Hello,  I am seeing the same behavior in version 0.109.0.  The issue is that the receiver never shuts down, it only hangs while trying.  Relevant log entry is:
`{"level":"info","ts":"2024-09-26T11:27:48.191-0400","caller":"kafkareceiver@v0.108.0/kafka_receiver.go:388","msg":"Consumer stopped","kind":"receiver","name":"kafka/kafkastream__logs","data_type":"logs","error":"context canceled"}`
The service needs to be manually stopped each time as it will hang here.

**Comment by djaglowski** [comment]
It looks like this might be a very slightly different problem but also may be an easy fix (See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35438). I'm just going to reopen this issue and resolve it again with the new PR if that works.

**Comment by djaglowski** [comment]
My quick fix was wishful thinking. Still, I given that we have two reports of the receiver _still_ not shutting down correctly after #32720, I think we might as well leave the issue open until we have a robust solution.

**Comment by jsirianni** [comment]
I was seeing this as well. The receiver is working great until it is time to shutdown.

**Comment by dpaasman00** [comment]
Working on a fix for this!

**Comment by dpaasman00** [comment]
Above PR should resolve this issue!
